Ensure each certificate has at least one supported key before encrypting

rocodes opened the pull request:
## Status

Ready for review (ready for initial review feedback - leaving in draft mode for that reason) 

## Description of Changes

Fixes #6895 

Changes proposed in this pull request:

- Before encryption, ensure each certificate contains at least one supported key (according to the specified policy, currently the [StandardPolicy](https://docs.sequoia-pgp.org/sequoia_openpgp/policy/struct.StandardPolicy.html)).

- Add test to ensure legacy (1024-bit RSA) key is rejected

## Testing

How should the reviewer test this PR?

- Visual review +/ code structure review
- Tests are passing and test cases are accounted for

## Deployment

n/a

## Checklist

### If you made changes to the server application code:

- [x] Linting (`make lint`) and tests (`make test`) pass in the development container **`make rust-lint`, `make rust-test`**

[...]

Choose one of the following:

- [ ] I have opened a PR in the [docs repo](https://github.com/freedomofpress/securedrop-docs) for these changes, or will do so later
- [ ] I would appreciate help with the documentation
- [x] These changes do not require documentation

rocodes committed:
Ensure each certificate has at least one supported key before encrypting

rocodes commented on redwood/src/lib.rs:
```diff
@@ -195,11 +209,35 @@ pub fn decrypt(
+        -----END PGP PUBLIC KEY BLOCK-----";
+    const BAD_KEY_FINGERPRINT: &str =
+        "56CED09CBE49E43E42D73D35E5159B8920CE5F31";
+
```
This seems...kinda ugly, so I'm open to other ways of doing it, but I wasn't sure the best way to include an unsupported key, since the CertBuilder doesn't have an option for unsupported CipherSuites and I got a bit into the weeds looking at how to generate certs that we don't actually want to support :)

legoktm replied:
I think just hardcoding the key is the right way to go, but I'd recommend just storing it in a file and then using the `include_str!()` macro (https://doc.rust-lang.org/std/macro.include_str.html) to set it as a constant. Plus that would make it simpler to share one bad SHA-1 key here and for your other admin+sq-keyring-linter PR.

rocodes commented on redwood/src/lib.rs:
```diff
@@ -217,6 +255,51 @@ mod tests {
+                actual_supported_keys += 1;
+            }
+            assert_eq!(expected_supported_keys, actual_supported_keys);
+        }
```
(It may be fine to just skip to testing `encrypt_message` as a whole, in which case this test function can be shorter n sweeter)

legoktm replied:
I would actually do the opposite, *just* test `encrypt_message` correctly fails instead of trying to count the number of supported keys, which requires duplicating the supported keys logic. If we want to test the underlying iterator then I think we should be splitting it out to a separate function.

legoktm commented on redwood/src/lib.rs:
```diff
@@ -131,18 +136,27 @@ fn encrypt(
+            .collect();
+
+        // Each certificate must have at least one supported encryption key
+        match keys.len() {
```
I think using an if/else statement here is clearer. IMO match is better when you're comparing enum types or have multiple if/elseif/elseif/else type blocks.

legoktm commented on redwood/src/lib.rs:
```diff
@@ -32,6 +34,9 @@ pub enum Error {
     NotUnicode(#[from] FromUtf8Error),
+    // Distinct from OpenPgp error because it's about our own logic
+    #[error("Encryption Error: {0}")]
+    Encryption(String),
```
In Rust the convention is to have a different type for each possible error. So instead of an `Encryption(String)` type with a message that describes the specific issue, we could just have `NoSupportedKeys`, and put the "At least one supported key is needed per certificate" from below into the error message itself. 

(Note that I would suggest doing `NoSupportedKeys(String)` so we can include the fingerprint, see below)

legoktm commented on redwood/src/lib.rs:
```diff
@@ -131,18 +136,27 @@ fn encrypt(
     // These filter options should be kept in sync with `Helper::decrypt()`.
     for cert in certs.iter() {
-        for key in cert
+        let keys: Vec<ValidErasedKeyAmalgamation<PublicParts>> = cert
```
I think Rust should be able to figure out the inner type, so in theory you should just need to write `let keys: Vec<_> = cert...collect()` (Nothing wrong with spelling out the full type if you want, I just find stuff like `ValidErasedKeyAmalgamation<PublicParts>` annoying to write so I try to skip them)

legoktm commented on redwood/src/lib.rs:
```diff
@@ -131,18 +136,27 @@ fn encrypt(
+        match keys.len() {
+            0 => {
+                return Err(Error::Encryption(
+                    "At least one supported key is needed per certificate"
```
Can we pass the cert's fingerprint here so for messages encrypted to multiple keys we can figure out which one is at fault?

legoktm reviewed (changes requested):
Functionally looks good, comments are mostly about Rust stuff.

rocodes committed:
Move test key and fingerprint into res/ directory.

Simplify testing of unsupported GPG key.

Update NoSupportedKeys error to display fingerprint of problem key in
error msg.

rocodes commented:
Thanks so much for all your helpful feedback @legoktm  - I think I've addressed everything in 5aff332

legoktm reviewed (approved):
Awesome!
